File: communication/rabbitmq/simple-demo/go-server/publisher/handler.go
package publisher

import (
	"context"
	"encoding/json"
	"net/http"
	"time"

	"github.com/labstack/echo/v4"
	amqp "github.com/rabbitmq/amqp091-go"

	"github.com/seungyeop-lee/poc/communication/rabbitmq/simple-demo/go-server/common"
	"github.com/seungyeop-lee/poc/communication/rabbitmq/simple-demo/go-server/config"
)

type Handler struct {
	rabbit *config.RabbitMQ
}

func NewHandler(rabbit *config.RabbitMQ) *Handler {
	return &Handler{rabbit: rabbit}
}

func (h *Handler) Apply(e *echo.Echo) {
	e.POST("/publish", h.PublishMessage)
}

func (h *Handler) PublishMessage(c echo.Context) error {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	payload := common.PayloadMessage{
		Name:        "from go server",
		CurrentTime: time.Now().UnixMilli(),
		Enable:      true,
		Data:        []byte("Hello from Go RabbitMQ!"),
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	err = h.rabbit.Channel.PublishWithContext(ctx,
		common.ExchangeName,
		"poc.communication.rabbitmq.go",
		false,
		false,
		amqp.Publishing{
			ContentType: "application/json",
			Body:        body,
		})
	if err != nil {
		return err
	}

	return c.NoContent(http.StatusOK)
}

File: communication/rabbitmq/simple-demo/python-server/python_server/publisher/handler.py
import base64
import json
import time

import aio_pika
from fastapi import APIRouter

from python_server.common.constants import EXCHANGE_NAME
from python_server.common.types import PayloadMessage
from python_server.config.rabbitmq import RabbitMQ

router = APIRouter()


class Handler:
    def __init__(self, rabbit: RabbitMQ):
        self.rabbit = rabbit
        self.router = router

    def setup_routes(self):
        self.router.add_api_route("/publish", self.publish_message, methods=["POST"])

    async def publish_message(self):
        payload = PayloadMessage(
            name="from python server",
            current_time=int(time.time() * 1000),
            enable=True,
            data=base64.b64encode(b"Hello from Python RabbitMQ!").decode('utf-8'),
        )

        exchange = await self.rabbit.channel.declare_exchange(
            EXCHANGE_NAME,
            aio_pika.ExchangeType.TOPIC,
            durable=True,
        )

        message = aio_pika.Message(
            body=json.dumps(payload.model_dump(by_alias=True)).encode(),
            content_type="application/json",
        )

        await exchange.publish(
            message,
            routing_key="poc.communication.rabbitmq.python",
        )

        return {"status": "ok"}